### ISSUE TITLE
Support Azure AD B2C

### ISSUE THREAD

**Issue by Narnain** [xamarin/Xamarin.Auth#105]
It would be great to add support for Azure AD B2C including the ability to sign in via social providers, and retrieve refresh and bearer tokens for communicating with a secured WebAPI endpoint.

**Comment by tylerstelzer** [comment]
If you derive a class from OAuth2Authenticator, you can make it work with B2C with a couple modifications/overrides.  The initial url will need to be overriden to include the signinpolicy as a query string parameter.  You will also need to change "access_token" to "id_token" as B2C decided to change it.  Hope that helps.

**Comment by gabog** [comment]
@tylerstelzer would you be able to provide more details on the overrides (or some sample code) I am trying this approach and can't get it to work. 
Thanks

**Comment by newky2k** [comment]
this is on the roadmap

**Comment by Robert-Altland** [comment]
Any update on this? I have subclassed the authenticator and can guess at what methods need to be overridden, but if you've already got it working, perhaps you can share? I'm not looking for full documentation, just something to get started.

**Comment by Robert-Altland** [comment]
OK, I'll provide details on how I did this, along with where I got blocked. Perhaps it'll be useful to someone here, and perhaps it will prompt some more detail from the author. 

I was able to complete the oauth implicit flow against our Azure B2C tenant out of the box, but I needed to subclass OAuth2Authenticator to get the authorization code flow to work. Implicit flow returns an access token but not a refresh token, and since the access token is typically short lived (1 hour usually), using the implicit flow would result in frequent reauth requests and a bad UX. I needed to complete oauth's authorization code flow. That flow returns a code from authentication that can then be exchanged for access and refresh tokens, enabling silent renewal of the access token when it expires. 

To do this I overrode OnRedirectPageLoaded, but that's probably not the best place to do it. Using a decompiler I was able to browse the source for Oauth2Authenticator to see what was happening there. It looks like there's code to handle this flow. (I expected this, since it's an oauth component and this is a flow it supports.) I think it was calling RequestAccessTokenAsync internally. Perhaps that's the source of the failure to complete the authorization code flow out of the box.

Anyway, regardless of whether or not this is the best place to customize this class, it's what I did and it works, and I don't see any direction from the author, so here are some details. If they're wrong enough, maybe it'll prompt a correction.

Inside OnRedirectPageLoaded, I look for the existence of a code querystring parameter, and if found, I use that to do a POST using regular old HttpClient. Once the response is received, I rejoin the Xamarin.Auth authentication flow by either calling OnError if an error was encountered, or OnRetrieveAccountProperties if things are good. Both trigger Authentication events, either Completed or Error. This is largely how the existing implementation works -- I just bypassed the broken RequestAccessTokenAsync (the method isn't broken, just how it constructs the request for Azure B2C that seems broken) and supplied my own using HttpClient to do the follow up POST to exchange my code for tokens. 

However, I'm not able to complete a password reset against an Azure B2C tenant. It seems like there's an ajax page refresh or something after you submit your email address. This refresh is supposed to provide a textbox for you to enter the code you receive in order to continue. That refresh never happens and I'm stuck on a form that I have no other choice but to cancel.
